

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Detect complex regions &mdash; odgi 3c4a8a4 documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Extract selected loci" href="extract_selected_loci.html" />
    <link rel="prev" title="Exploratory analysis" href="exploratory_analysis.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> odgi
          

          
          </a>

          
            
            
              <div class="version">
                v0.5.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../index.html">Welcome</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quick_start.html">Quick Start</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="exploratory_analysis.html">Exploratory analysis</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Detect complex regions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#synopsis">Synopsis</a></li>
<li class="toctree-l3"><a class="reference internal" href="#steps">Steps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#get-the-human-chr8-dataset">Get the Human chr8 dataset</a></li>
<li class="toctree-l4"><a class="reference internal" href="#obtain-the-depth-over-the-pangenome">Obtain the depth over the pangenome</a></li>
<li class="toctree-l4"><a class="reference internal" href="#plot-the-depth">Plot the depth</a></li>
<li class="toctree-l4"><a class="reference internal" href="#the-centromere-s-higher-order-repeat-hor-organization">The centromere’s Higher-Order Repeat (HOR) organization</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="extract_selected_loci.html">Extract selected <cite>loci</cite></a></li>
<li class="toctree-l2"><a class="reference internal" href="sorting_layouting.html">Sorting and Layouting</a></li>
<li class="toctree-l2"><a class="reference internal" href="translate_path_positions_between_graphs.html">Translate path positions between graphs</a></li>
<li class="toctree-l2"><a class="reference internal" href="remove_artifacts_and_complex_regions.html">Remove artifacts and complex regions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../commands.html">Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="../binding.html">Python Binding</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">odgi</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../tutorials.html">Tutorials</a> &raquo;</li>
        
      <li>Detect complex regions</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../_sources/rst/tutorials/detect_complex_regions.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="detect-complex-regions">
<h1>Detect complex regions<a class="headerlink" href="#detect-complex-regions" title="Permalink to this headline">¶</a></h1>
<div class="section" id="synopsis">
<h2>Synopsis<a class="headerlink" href="#synopsis" title="Permalink to this headline">¶</a></h2>
<p>Human chromosomes have large regions of highly identical repeats. They can be found in as clusters centromeres,
regions of segmental duplication, or in the acrocentric short arms of chromosomes.
<a class="reference external" href="https://doi.org/10.1038/s41586-021-03420-7">Logsdon et al., Nature 2021</a> showed that chromosome 8 carries a modestly sized
centromere of approximately 1.5–2.2 Mb, in which AT-rich, 171-base-pair (bp) α-satellite
repeats are organized into a well-defined higher-order repeat (HOR) array. <code class="docutils literal notranslate"><span class="pre">odgi</span></code> offers tools to detect and explore
such regions.</p>
</div>
<div class="section" id="steps">
<h2>Steps<a class="headerlink" href="#steps" title="Permalink to this headline">¶</a></h2>
<div class="section" id="get-the-human-chr8-dataset">
<h3>Get the Human chr8 dataset<a class="headerlink" href="#get-the-human-chr8-dataset" title="Permalink to this headline">¶</a></h3>
<p>Download the pangenome graph of the <a class="reference external" href="https://s3-us-west-2.amazonaws.com/human-pangenomics/pangenomes/scratch/2021_05_06_pggb/gfas/chr8.pan.gfa.gz">Human chromosome 8</a>
in <code class="docutils literal notranslate"><span class="pre">GFA</span></code> format, decompress it, and convert it to a graph in <code class="docutils literal notranslate"><span class="pre">odgi</span></code> format:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget -c https://s3-us-west-2.amazonaws.com/human-pangenomics/pangenomes/scratch/2021_05_06_pggb/gfas/chr8.pan.gfa.gz
gunzip chr8.pan.gfa.gz

odgi build -g chr8.pan.gfa -o chr8.pan.og --threads <span class="m">2</span> -P
</pre></div>
</div>
<p>The last command creates a file called <code class="docutils literal notranslate"><span class="pre">chr8.pan.og</span></code>, which contains the input graph in <code class="docutils literal notranslate"><span class="pre">odgi</span></code> format. This graph contains
the contigs of 88 haploid, phased human genome assemblies from 44 individuals, plus the chm13 and GRCh38 reference genomes.</p>
</div>
<div class="section" id="obtain-the-depth-over-the-pangenome">
<h3>Obtain the depth over the pangenome<a class="headerlink" href="#obtain-the-depth-over-the-pangenome" title="Permalink to this headline">¶</a></h3>
<p>To identify the repetitive sequences, we can analyze the <strong>depth</strong> in the graph. Here we define as <strong>node depth</strong> the
number of times in which the node is crossed by all the paths present in the graph.</p>
<p>To obtain the depth over the pangenome graph, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi depth -i chr8.pan.og -r chm13#chr8 <span class="p">|</span> <span class="se">\</span>
    bedtools makewindows -b /dev/stdin -w <span class="m">5000</span> &gt; chm13.chr8.w5kbps.bed

odgi depth -i chr8.pan.og -b chm13.chr8.w5kbps.bed --threads <span class="m">2</span> <span class="p">|</span> <span class="se">\</span>
    bedtools sort &gt; chr8.pan.depth.w5kbps.bed
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">chm13.chr8.w5kbps.bed</span></code> file contains 5000 bps interval windows across the chm13 reference chromosome 8. The depth
is computed for each of these windows, writing the result in the <code class="docutils literal notranslate"><span class="pre">chr8.pan.depth.w5kbps.bed</span></code> file, in <code class="docutils literal notranslate"><span class="pre">BED</span></code> format.
<code class="docutils literal notranslate"><span class="pre">-r</span></code> specifies the path name from which to compute the depth from. <code class="docutils literal notranslate"><span class="pre">-b</span></code> specifies the BED ranges of which the
depths should be calculated of.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>It will take a while to obtain the depth over all the windows. You can find the gzipped result in the <code class="docutils literal notranslate"><span class="pre">test</span></code> folder
in the root of the <code class="docutils literal notranslate"><span class="pre">odgi</span></code> project, in the <code class="docutils literal notranslate"><span class="pre">chr8.pan.depth.w5kbps.bed.gz</span></code> file. Gunzip the file before using it.</p>
</div>
<p>Take a look at the first rows of the <code class="docutils literal notranslate"><span class="pre">chr8.pan.depth.w5kbps.bed</span></code> file:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head chr8.pan.depth.w5kbps.bed -n <span class="m">5</span> <span class="p">|</span> column -t
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>chm13#chr8  0      5000   60.1014
chm13#chr8  5000   10000  89.5402
chm13#chr8  10000  15000  91.1302
chm13#chr8  15000  20000  90.0506
chm13#chr8  20000  25000  94.0238
</pre></div>
</div>
<p>The 4-th column indicates the <code class="docutils literal notranslate"><span class="pre">mean</span> <span class="pre">depth</span></code> of each window. For each node within a window, the product between the node
depth and its length (in bp) is calculated. The overall total sum is divided by the length of the window (in bp). This
results in the <code class="docutils literal notranslate"><span class="pre">mean</span> <span class="pre">depth</span></code>.</p>
</div>
<div class="section" id="plot-the-depth">
<h3>Plot the depth<a class="headerlink" href="#plot-the-depth" title="Permalink to this headline">¶</a></h3>
<p>Plotting the <code class="docutils literal notranslate"><span class="pre">mean</span> <span class="pre">depth</span></code> distribution:</p>
<img alt="../../_images/chr8.mean_depth.png" src="../../_images/chr8.mean_depth.png" />
<p><code class="docutils literal notranslate"><span class="pre">Mean</span> <span class="pre">depth</span></code> distribution of chromosome 8: Three high depth regions are present. <em>Left:</em> The β-defensin gene
cluster. <em>Mid:</em> The centromere. <em>Right:</em> The neo-centromere.</p>
</div>
<div class="section" id="the-centromere-s-higher-order-repeat-hor-organization">
<h3>The centromere’s Higher-Order Repeat (HOR) organization<a class="headerlink" href="#the-centromere-s-higher-order-repeat-hor-organization" title="Permalink to this headline">¶</a></h3>
<p>The chromosome 8 centromeric HOR array is primarily composed of four distinct HOR types represented by 4, 7, 8 or
11 α-satellite monomer cassettes. Thankfully, <a class="reference external" href="https://doi.org/10.1038/s41586-021-03420-7">Logsdon et al., Nature 2021</a> provide HOR
annotations (you can find them in the <code class="docutils literal notranslate"><span class="pre">test</span></code> folder in the root of the <code class="docutils literal notranslate"><span class="pre">odgi</span></code> project, in the <code class="docutils literal notranslate"><span class="pre">chr8_HORarray_organization.colors.bed</span></code>
file), so we can take a closer look at the centromere:</p>
<img alt="../../_images/chr8_HOR.mean_depth.2.png" src="../../_images/chr8_HOR.mean_depth.2.png" />
<p><code class="docutils literal notranslate"><span class="pre">Mean</span> <span class="pre">depth</span></code> of the chromosome 8 centromeric HOR array. Every letter indicates an alpha-satellite monomer in the HOR (so, <code class="docutils literal notranslate"><span class="pre">A,B,C,D,E,F,G,H,I,J,K</span></code>
would indicate an HOR with 11 alpha-satellite monomers). The <code class="docutils literal notranslate"><span class="pre">mean</span> <span class="pre">depth</span></code> drop falls into the hypomethylated and CENP-A-enriched regions, that have the highest consistent
entropy in the entire array. This is in concordance with the <a class="reference external" href="https://doi.org/10.1038/s41586-021-03420-7">Logsdon et al., Nature 2021</a>
publication, see the figure below. The drop in the depth may indicate hypervariability between genomes in the transition from methylated
to unmethylated DNA, or that the region is not alignable.</p>
<img alt="../../_images/chr8_HOR.organization.2.png" src="../../_images/chr8_HOR.organization.2.png" />
<p>Overview of the chm13 chromosome 8 D8Z2 α-satellite HOR array. <strong>a (top):</strong> Monomer arrangements of the HOR. <strong>a (bottom):</strong>
The entropy plot of chromosome 8 reveals that the hypomethylated and CENP-A-enriched regions, that have the highest consistent
entropy in the entire array. <strong>b:</strong> Structure of the α-satellite monomers listed by HOR. <strong>c:</strong> Frequency plot of the α-satellite monomers in the HOR.
Figure taken from <a class="reference external" href="https://www.nature.com/articles/s41586-021-03420-7/figures/12">Logsdon et al., Nature 2021</a>.</p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="extract_selected_loci.html" class="btn btn-neutral float-right" title="Extract selected loci" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="exploratory_analysis.html" class="btn btn-neutral float-left" title="Exploratory analysis" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Erik Garrison. Revision v0.5.1-3c4a8a4.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>