

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Extract selected loci &mdash; odgi 3c4a8a4 documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Sorting and Layouting" href="sorting_layouting.html" />
    <link rel="prev" title="Detect complex regions" href="detect_complex_regions.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> odgi
          

          
          </a>

          
            
            
              <div class="version">
                v0.5.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../index.html">Welcome</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quick_start.html">Quick Start</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="exploratory_analysis.html">Exploratory analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="detect_complex_regions.html">Detect complex regions</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Extract selected <cite>loci</cite></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#synopsis">Synopsis</a></li>
<li class="toctree-l3"><a class="reference internal" href="#steps">Steps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#build-the-lipoprotein-a-graph">Build the Lipoprotein A graph</a></li>
<li class="toctree-l4"><a class="reference internal" href="#display-path-names">Display path names</a></li>
<li class="toctree-l4"><a class="reference internal" href="#obtain-the-lipoprotein-a-variants">Obtain the Lipoprotein A variants</a></li>
<li class="toctree-l4"><a class="reference internal" href="#extract-a-sub-graph-with-a-variant-inside">Extract a sub-graph with a variant inside</a></li>
<li class="toctree-l4"><a class="reference internal" href="#visualize-the-sub-graph">Visualize the sub-graph</a></li>
<li class="toctree-l4"><a class="reference internal" href="#get-the-human-chr6-dataset">Get the Human chr6 dataset</a></li>
<li class="toctree-l4"><a class="reference internal" href="#extract-the-mhc-locus">Extract the MHC <cite>locus</cite></a></li>
<li class="toctree-l4"><a class="reference internal" href="#extract-a-sub-graph-with-the-hla-genes">Extract a sub-graph with the HLA genes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id1">Visualize the sub-graph</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="sorting_layouting.html">Sorting and Layouting</a></li>
<li class="toctree-l2"><a class="reference internal" href="translate_path_positions_between_graphs.html">Translate path positions between graphs</a></li>
<li class="toctree-l2"><a class="reference internal" href="remove_artifacts_and_complex_regions.html">Remove artifacts and complex regions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../commands.html">Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="../binding.html">Python Binding</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">odgi</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../tutorials.html">Tutorials</a> &raquo;</li>
        
      <li>Extract selected <cite>loci</cite></li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../_sources/rst/tutorials/extract_selected_loci.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="extract-selected-loci">
<h1>Extract selected <cite>loci</cite><a class="headerlink" href="#extract-selected-loci" title="Permalink to this headline">¶</a></h1>
<div class="section" id="synopsis">
<h2>Synopsis<a class="headerlink" href="#synopsis" title="Permalink to this headline">¶</a></h2>
<p>Pangenome graphs model the full set of genomic elements in a gives species or clade. Nevertheless, downstream analyses
may require focusing on specific pangenomic regions. <code class="docutils literal notranslate"><span class="pre">odgi</span></code> allows to extract <cite>loci</cite> of interest from the pangenome graph,
resulting in sub-graphs of the full pangenome graph. It is demonstrated that one can work with such sub-graphs as easily as
handling the full ones.</p>
</div>
<div class="section" id="steps">
<h2>Steps<a class="headerlink" href="#steps" title="Permalink to this headline">¶</a></h2>
<div class="section" id="build-the-lipoprotein-a-graph">
<h3>Build the Lipoprotein A graph<a class="headerlink" href="#build-the-lipoprotein-a-graph" title="Permalink to this headline">¶</a></h3>
<p>Assuming that your current working directory is the root of the <code class="docutils literal notranslate"><span class="pre">odgi</span></code> project, to construct an <code class="docutils literal notranslate"><span class="pre">odgi</span></code> file from the
<code class="docutils literal notranslate"><span class="pre">LPA</span></code> dataset in <code class="docutils literal notranslate"><span class="pre">GFA</span></code> format, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi build -g test/LPA.gfa -o LPA.og
</pre></div>
</div>
<p>The command creates a file called <code class="docutils literal notranslate"><span class="pre">LPA.og</span></code>, which contains the input graph in <code class="docutils literal notranslate"><span class="pre">odgi</span></code> format. This graph contains
13 contigs from 7 haploid human genome assemblies from 6 individuals plus the chm13 cell line. The contigs cover the
<a class="reference external" href="https://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000198670">Lipoprotein A (LPA)</a> <cite>locus</cite>, which encodes the
Apo(a) protein.</p>
</div>
<div class="section" id="display-path-names">
<h3>Display path names<a class="headerlink" href="#display-path-names" title="Permalink to this headline">¶</a></h3>
<p>The path’s names are:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi paths -i LPA.og -L
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>chm13__LPA__tig00000001
HG002__LPA__tig00000001
HG002__LPA__tig00000005
HG00733__LPA__tig00000001
HG00733__LPA__tig00000008
HG01358__LPA__tig00000002
HG01358__LPA__tig00000010
HG02572__LPA__tig00000005
HG02572__LPA__tig00000001
NA19239__LPA__tig00000002
NA19239__LPA__tig00000006
NA19240__LPA__tig00000001
NA19240__LPA__tig00000012
</pre></div>
</div>
</div>
<div class="section" id="obtain-the-lipoprotein-a-variants">
<h3>Obtain the Lipoprotein A variants<a class="headerlink" href="#obtain-the-lipoprotein-a-variants" title="Permalink to this headline">¶</a></h3>
<p>To see the variants for the two contigs of the <code class="docutils literal notranslate"><span class="pre">HG02572</span></code> sample, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>zgrep -v <span class="s1">&#39;^##&#39;</span> test/LPA.chm13__LPA__tig00000001.vcf.gz <span class="p">|</span> head -n <span class="m">9</span> <span class="p">|</span> cut -f <span class="m">1</span>-9,16,17 <span class="p">|</span> column -t
</pre></div>
</div>
<p>The pangenome graphs embed the full mutual relationships between the embedded genomes and their variation. In the following example,
the variants were called with respect to the <code class="docutils literal notranslate"><span class="pre">chm13__LPA__tig00000001</span></code> contig, which was used as the reference path.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>#CHROM                   POS   ID      REF  ALT  QUAL  FILTER  INFO                        FORMAT  HG02572__LPA__tig00000001  HG02572__LPA__tig00000005
chm13__LPA__tig00000001  6     &gt;3&gt;6    T    A    60    .       AC=1;AF=0.5;AN=2;LV=0;NS=2  GT      1                          0
chm13__LPA__tig00000001  134   &gt;6&gt;8    G    GT   60    .       AC=1;AF=0.5;AN=2;LV=0;NS=2  GT      1                          0
chm13__LPA__tig00000001  156   &gt;8&gt;11   A    G    60    .       AC=1;AF=0.5;AN=2;LV=0;NS=2  GT      0                          1
chm13__LPA__tig00000001  252   &gt;11&gt;13  GT   G    60    .       AC=2;AF=1;AN=2;LV=0;NS=2    GT      1                          1
chm13__LPA__tig00000001  265   &gt;13&gt;16  T    A    60    .       AC=2;AF=1;AN=2;LV=0;NS=2    GT      1                          1
chm13__LPA__tig00000001  371   &gt;16&gt;18  AT   A    60    .       AC=2;AF=1;AN=2;LV=0;NS=2    GT      1                          1
chm13__LPA__tig00000001  996   &gt;18&gt;21  AG   ATT  60    .       AC=2;AF=1;AN=2;LV=0;NS=2    GT      1                          1
chm13__LPA__tig00000001  1050  &gt;21&gt;23  G    GT   60    .       AC=1;AF=0.5;AN=2;LV=0;NS=2  GT      1                          0
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">ID</span></code> field in the <a class="reference external" href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">VCF</a> lists the nodes involved in the variant. A <code class="docutils literal notranslate"><span class="pre">&gt;</span></code> means that the node is visited in forward
orientation, a <code class="docutils literal notranslate"><span class="pre">&lt;</span></code> means that the node is visited in reverse orientation.</p>
</div>
<div class="section" id="extract-a-sub-graph-with-a-variant-inside">
<h3>Extract a sub-graph with a variant inside<a class="headerlink" href="#extract-a-sub-graph-with-a-variant-inside" title="Permalink to this headline">¶</a></h3>
<p>The insertion at position <strong>1050</strong> (<em>G &gt; GT</em>, the last line in the VCF snippet above) is present only in one of the  <code class="docutils literal notranslate"><span class="pre">HG02572</span></code>’s contig (<code class="docutils literal notranslate"><span class="pre">HG02572__LPA__tig00000001</span></code>).
To extract the sub-graph where this insertion falls, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi extract -i LPA.og -n <span class="m">23</span> -c <span class="m">1</span> -o LPA.21_23_G_GT.og
</pre></div>
</div>
<p>The instruction extracts:</p>
<ul class="simple">
<li><p>The node with <strong>ID 23</strong> (<code class="docutils literal notranslate"><span class="pre">-n</span> <span class="pre">23</span></code>),</p></li>
<li><p>the nodes reachable from this node following a single edge (<code class="docutils literal notranslate"><span class="pre">-c</span> <span class="pre">1</span></code>) in the graph topology,</p></li>
<li><p>the edges connecting all the extracted nodes, and</p></li>
<li><p>the paths traversing all the extracted nodes.</p></li>
</ul>
<p>To have basic information on the sub-graph, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi stats -i LPA.21_23_G_GT.og -S
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>#length nodes   edges   paths
644     5       6       3
</pre></div>
</div>
<p>The extracted path’s names are:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi paths -i LPA.21_23_G_GT.og -L
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>chm13__LPA__tig00000001:997-1640
HG02572__LPA__tig00000005:999-1641
HG02572__LPA__tig00000001:1035-1678
</pre></div>
</div>
<p>The sub-graph contains the contig used as a reference in the <code class="docutils literal notranslate"><span class="pre">VCF</span></code> file, and the two <code class="docutils literal notranslate"><span class="pre">HG02572</span></code>’s contigs.</p>
</div>
<div class="section" id="visualize-the-sub-graph">
<h3>Visualize the sub-graph<a class="headerlink" href="#visualize-the-sub-graph" title="Permalink to this headline">¶</a></h3>
<p>To visualize the sub-graph, we can also use external tools as <a class="reference external" href="https://github.com/rrwick/Bandage">Bandage</a>, which
supports grpahs in <code class="docutils literal notranslate"><span class="pre">GFA</span></code> format. To covert the graph in <code class="docutils literal notranslate"><span class="pre">odgi</span></code> format in a graph in <code class="docutils literal notranslate"><span class="pre">GFA</span></code> format, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi view -i LPA.21_23_G_GT.og -g &gt; LPA.21_23_G_GT.gfa
</pre></div>
</div>
<p>Then, open the <code class="docutils literal notranslate"><span class="pre">LPA.21_23_G_GT.gfa</span></code> file with <code class="docutils literal notranslate"><span class="pre">Bandage</span></code>.</p>
<img alt="../../_images/LPA.21_23_G_GT.png" src="../../_images/LPA.21_23_G_GT.png" />
<p>The image shows the graph topology, where each colored rectangle represents a node. In particular, three paths support
nodes with <strong>ID 21</strong> and <strong>23</strong>, and only one path supports the node with <strong>ID 22</strong>. The node with <strong>ID 22</strong> represents in the graph the
additional nucleotide <code class="docutils literal notranslate"><span class="pre">T</span></code> presents in the <code class="docutils literal notranslate"><span class="pre">HG02572__LPA__tig00000001</span></code> contig as an insertion.</p>
</div>
<div class="section" id="get-the-human-chr6-dataset">
<h3>Get the Human chr6 dataset<a class="headerlink" href="#get-the-human-chr6-dataset" title="Permalink to this headline">¶</a></h3>
<p>Download the pangenome graph of the <a class="reference external" href="https://s3-us-west-2.amazonaws.com/human-pangenomics/pangenomes/scratch/2021_05_06_pggb/gfas/chr6.pan.gfa.gz">Human chromosome 6</a>
in <code class="docutils literal notranslate"><span class="pre">GFA</span></code> format, decompress it, and convert it to a graph in <code class="docutils literal notranslate"><span class="pre">odgi</span></code> format:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget -c https://s3-us-west-2.amazonaws.com/human-pangenomics/pangenomes/scratch/2021_05_06_pggb/gfas/chr6.pan.gfa.gz
gunzip chr6.pan.gfa.gz

odgi build -g chr6.pan.gfa -o chr6.pan.og --threads <span class="m">2</span> -P
</pre></div>
</div>
<p>The last command creates a file called <code class="docutils literal notranslate"><span class="pre">chr6.pan.og</span></code>, which contains the input graph in <code class="docutils literal notranslate"><span class="pre">odgi</span></code> format. This graph contains contigs of
88 haploid, phased human genome assemblies from 44 individuals, plus the chm13 and GRCh38 reference genomes.</p>
</div>
<div class="section" id="extract-the-mhc-locus">
<h3>Extract the MHC <cite>locus</cite><a class="headerlink" href="#extract-the-mhc-locus" title="Permalink to this headline">¶</a></h3>
<p>The <a class="reference external" href="https://en.wikipedia.org/wiki/Major_histocompatibility_complex">major histocompatibility complex</a> (MHC) is a large
<cite>locus</cite> in vertebrate DNA containing a set of closely linked polymorphic genes that code for cell surface proteins essential
for the adaptive immune system. In humans, the MHC region occurs on chromosome 6. The human MHC is also called the HLA
(human leukocyte antigen) complex (often just the HLA).</p>
<p>Assuming that your current working directory is the root of the <code class="docutils literal notranslate"><span class="pre">odgi</span></code> project, to see the coordinates of some HLA genes,
execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head test/chr6.HLA_genes.bed -n <span class="m">5</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>grch38#chr6     29722775        29738528        HLA-F
grch38#chr6     29826967        29831125        HLA-G
grch38#chr6     29941260        29945884        HLA-A
grch38#chr6     30489509        30494194        HLA-E
grch38#chr6     31268749        31272130        HLA-C
</pre></div>
</div>
<p>The coordinates are expressed with respect to the GRCh38 reference genome.</p>
</div>
<div class="section" id="extract-a-sub-graph-with-the-hla-genes">
<h3>Extract a sub-graph with the HLA genes<a class="headerlink" href="#extract-a-sub-graph-with-the-hla-genes" title="Permalink to this headline">¶</a></h3>
<p>To extract the sub-graph containing all the HLA genes annotated in the <code class="docutils literal notranslate"><span class="pre">chr6.HLA_genes.bed</span></code> file, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi extract -i chr6.pan.og -o chr6.pan.MHC.og -b chr6.HLA_genes.bed  -E --threads <span class="m">2</span> -P
</pre></div>
</div>
<p>The instruction extracts:</p>
<ul class="simple">
<li><p>The nodes belonging to the <code class="docutils literal notranslate"><span class="pre">grch38#chr6</span></code> path ranges specified in the the <code class="docutils literal notranslate"><span class="pre">chr6.HLA_genes.bed</span></code> file via <code class="docutils literal notranslate"><span class="pre">-b</span></code>,</p></li>
<li><p>all nodes between the min and max positions touched by the given path ranges, also if they belong to other paths (<code class="docutils literal notranslate"><span class="pre">-E</span></code>),</p></li>
<li><p>the edges connecting all the extracted nodes, and</p></li>
<li><p>the paths traversing all the extracted nodes.</p></li>
</ul>
<p>To have basic information on the sub-graph, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi stats -i chr6.pan.MHC.og -S
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>#length     nodes   edges   paths
3896981     216352  297890  97
</pre></div>
</div>
<p>There are 97 paths in the sub-graph. This means that for few individuals, more than one contig covers the MHC <cite>locus</cite>.</p>
</div>
<div class="section" id="id1">
<h3>Visualize the sub-graph<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p>To visualize the sub-graph with <code class="docutils literal notranslate"><span class="pre">odgi</span></code>, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi sort -i chr6.pan.MHC.og -o - -O <span class="p">|</span> <span class="se">\</span>
    odgi viz -i - -o chr6.pan.MHC.png -s <span class="s1">&#39;#&#39;</span> -P <span class="m">20</span>
</pre></div>
</div>
<p>To obtain the following PNG image:</p>
<img alt="../../_images/chr6.pan.MHC.png" src="../../_images/chr6.pan.MHC.png" />
<p>In this 1-Dimensional visualization all contigs of the same haplotype are represented with the same color (<code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">'#'</span></code>).
The majority of the haplotypes has one contig covering the whole <cite>locus</cite>, meanwhile in few of them, the <cite>locus</cite> is split
in several contigs. We had to apply <a class="reference internal" href="../commands/odgi_sort.html#odgi-sort"><span class="std std-ref">odgi sort</span></a> here in order to optimize (<code class="docutils literal notranslate"><span class="pre">-O</span></code>) our sub-graph. This ensures that
the node identifier space is compacted from one to the number of nodes in the sub-graph.</p>
<p>To see the haplotypes sorted by the number of contigs covering the MHC <cite>locus</cite>, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi paths -i chr6.pan.MHC.og -L <span class="p">|</span> cut -f <span class="m">1</span>,2 -d <span class="s1">&#39;#&#39;</span> <span class="p">|</span> uniq -c <span class="p">|</span> sort -k 1nr <span class="p">|</span> head
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>3 HG00733#2
2 HG00673#2
2 HG01071#1
2 HG02630#2
2 HG02818#2
2 HG03516#2
1 chm13#chr6:29595119-32911317
1 grch38#chr6:29722774-33089734
1 HG00438#1
1 HG00438#2
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="sorting_layouting.html" class="btn btn-neutral float-right" title="Sorting and Layouting" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="detect_complex_regions.html" class="btn btn-neutral float-left" title="Detect complex regions" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Erik Garrison. Revision v0.5.1-3c4a8a4.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>