

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>NAME &mdash; odgi e809b89 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> odgi
          

          
          </a>

          
            
            
              <div class="version">
                v0.5.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Welcome</a></li>
<li class="toctree-l1"><a class="reference internal" href="../rst/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../rst/quick_start.html">Quick Start</a></li>
<li class="toctree-l1"><a class="reference internal" href="../rst/tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../rst/commands.html">Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="../rst/binding.html">Python Binding</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">odgi</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>NAME</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/asciidocs/odgi.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="name">
<h1>NAME<a class="headerlink" href="#name" title="Permalink to this headline">¶</a></h1>
<p>odgi - dynamic succinct variation graph tool</p>
</div>
<div class="section" id="synopsis">
<h1>SYNOPSIS<a class="headerlink" href="#synopsis" title="Permalink to this headline">¶</a></h1>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_build.adoc#_odgi_build1">build</a> -g graph.gfa -o
graph.og</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_validate.adoc#_odgi_validate1">validate</a> -i graph.og</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_stats.adoc#_odgi_stats1">stats</a> -i graph.og -S</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_degree.adoc#_odgi_degree1">stats</a> -i graph.og -S</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_depth.adoc#_odgi_depth1">depth</a> -i graph.og</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_overlap.adoc#_odgi_overlap1">overlap</a> -i graph.og -r
path_name</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_cover.adoc#_odgi_cover1">cover</a> -i graph.og -o
graph.paths.og</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_extract.adoc#_odgi_extract1">extract</a> -i graph.og -p
prefix -r path_name:0-17</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_explode.adoc#_odgi_explode1">explode</a> -i graph.og -p
prefix</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_squeeze.adoc#_odgi_squeeze1">squeeze</a> -f
input_graphs.txt -o graphs.og</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_position.adoc#_odgi_position1">position</a> -i
target_graph.og -g</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_sort.adoc#_odgi_sort1">sort</a> -i graph.og -o
graph.sorted.og -p bSnSnS</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_view.adoc#_odgi_view1">view</a> -i graph.og -g</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_kmers.adoc#_odgi_kmers1">kmers</a> -i graph.og -c -k 23
-e 34 -D 50</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_unitig.adoc#_odgi_unitig1">unitig</a> -i graph.og -f -t
1324 -l 120</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_viz.adoc#_odgi_viz1">viz</a> -i graph.og -o graph.og.png
-x 1920 -y 1080 -R -t 28</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_draw.adoc#_odgi_draw1">draw</a> -i graph.og -c
coords.lay -p .png -x 1920 -y 1080 -R -t 28</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_paths.adoc#_odgi_paths1">paths</a> -i graph.og -f</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_prune.adoc#_odgi_prune1">prune</a> -i graph.og -o
graph.pruned.og -c 3 -C 345 -T</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_unchop.adoc#_odgi_unchop1">unchop</a> -i graph.og -o
graph.unchopped.og</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_normalize.adoc#_odgi_normalize1">normalize</a> -i
graph.og -o graph.normalized.og -I 100 -d</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_bin.adoc#_odgi_bin1">bin</a> -i graph.og -j -w 100 -s -g</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_matrix.adoc#_odgi_matrix1">matrix</a> -i graph.og -e -d</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_chop.adoc#_odgi_chop1">chop</a> -i graph.og -o
graph.choped.og -c 1000</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_groom.adoc#_odgi_groom1">groom</a> -i graph.og -o
graph.groomed.og</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_layout.adoc#_odgi_layout1">layout</a> -i graph.og -o
graph.svg -R 10 -m 100</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_break.adoc#_odgi_break1">break</a> -i graph.og -o
graph.broken.og -s 100 -d</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_pathindex.adoc#_odgi_pathindex1">pathindex</a> -i
graph.og -o graph.xp</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_panpos.adoc#_odgi_panpos1">panpos</a> -i graph.og -p
Chr1 -n 4</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_server.adoc#_odgi_server1">server</a> -i graph.og -p
4000 -ip 192.168.8.9</p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_test.adoc#_odgi_test1">test</a></p>
<p><strong>odgi</strong> <a class="reference external" href="#odgi_version.adoc#_odgi_version1">version</a></p>
</div>
<div class="section" id="description">
<h1>DESCRIPTION<a class="headerlink" href="#description" title="Permalink to this headline">¶</a></h1>
<p><strong>odgi</strong>, the <strong>Optimized Dynamic (genome) Graph Interface</strong>, links a
thrifty dynamic in-memory variation graph data model to a set of
algorithms designed for scalable sorting, pruning, transformation, and
visualization of very large <a class="reference external" href="https://pangenome.github.io/">genome
graphs</a>. <strong>odgi</strong> includes <a class="reference external" href="https://pangenome.github.io/odgi/odgipy.html">python
bindings</a> that can be
used to <a class="reference external" href="https://odgi.readthedocs.io/en/latest/rst/tutorial.html">directly interface with its data
model</a>. This
<strong>odgi</strong> manual provides detailed information about its features and
subcommands, including examples.</p>
</div>
<div class="section" id="commands">
<h1>COMMANDS<a class="headerlink" href="#commands" title="Permalink to this headline">¶</a></h1>
<p>Each command has its own man page which can be viewed using e.g. <strong>man
odgi_build.1</strong>. Below we have a brief summary of syntax and subcommand
description.</p>
<div class="line-block">
<div class="line"><strong>odgi build</strong> [<strong>-g, –gfa</strong>=<em>FILE</em>] [<strong>-o, –out</strong>=<em>FILE</em>]
[<em>OPTION</em>]…</div>
<div class="line">The odgi build(1) command constructs a succinct variation graph from a
GFA. Currently, only GFA1 is supported. For details of the format
please see <a class="reference external" href="https://github.com/GFA-spec/GFA-spec/blob/master/GFA1.md">https://github.com/GFA-spec/GFA-spec/blob/master/GFA1.md</a>.</div>
</div>
<p><strong>odgi validate</strong> [<strong>-i, –input</strong>=<em>FILE</em>] [<em>OPTION</em>]… The odgi
validate(1) command validates the graph (currently, it checks if the
paths are consistent with the graph topology).</p>
<div class="line-block">
<div class="line"><strong>odgi stats</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<em>OPTION</em>]…</div>
<div class="line">The odgi stats(1) command produces statistics of a variation graph.
Among other metrics, it can calculate the #nodes, #edges, #paths and
the total nucleotide length of the graph. Various histogram summary
options complement the tool. If [<strong>-B, –bed-multicov</strong>=<em>BED</em>] is
set, the metrics will be produced for the intervals specified in the
BED.</div>
</div>
<p><strong>odgi degree</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<em>OPTION</em>]… The odgi degree(1)
command describes the graph in terms of node degree. For the input
graph, it shows the node.count, edge.count, avg.degree, min.degree, and
max.degree.</p>
<p><strong>odgi depth</strong> [<strong>-i, –input</strong>=<em>FILE</em>] [<em>OPTION</em>]… The odgi depth(1)
command finds the depth of graph as defined by query criteria.</p>
<p><strong>odgi overlap</strong> [<strong>-i, –input</strong>=<em>FILE</em>] [<em>OPTION</em>]… The odgi
overlap(1) command finds the paths touched by the input paths.</p>
<div class="line-block">
<div class="line"><strong>odgi cover</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<strong>-o, –out</strong>=<em>FILE</em>]
[<em>OPTION</em>]…</div>
<div class="line">The odgi cover(1) command finds a path cover of a variation graph,
with a specified number of paths per component.</div>
</div>
<p><strong>odgi extract</strong> [<strong>-f, –input-graphs</strong>=<em>FILE</em>] [<strong>-o,
–out</strong>=<em>FILE</em>] [<em>OPTION</em>]… The odgi extract(1) command extracts
parts of the graph as defined by query criteria.</p>
<div class="line-block">
<div class="line"><strong>odgi explode</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<strong>-p,
–prefix</strong>=<em>STRING</em>] [<em>OPTION</em>]…</div>
<div class="line">The odgi explode(1) command breaks a graph into connected components,
writing each component in its own file.</div>
</div>
<p><strong>odgi squeeze</strong> [<strong>-f, –input-graphs</strong>=<em>FILE</em>] [<strong>-o,
–out</strong>=<em>FILE</em>] [<em>OPTION</em>]… The odgi squeeze(1) command squeezes
multiple graphs into the same file.</p>
<p><strong>odgi position</strong> [<strong>-i, –target</strong>=<em>FILE</em>] [<em>OPTION</em>]… The odgi
position(1) command position parts of the graph as defined by query
criteria.</p>
<div class="line-block">
<div class="line"><strong>odgi sort</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<strong>-o, –out</strong>=<em>FILE</em>]
[<em>OPTION</em>]…</div>
<div class="line">The odgi sort(1) command sorts a succinct variation graph. The command
offers a diverse palette of sorting algorithms to determine the node
order:</div>
</div>
<ul class="simple">
<li><p>A topological sort: A graph can be sorted via <a class="reference external" href="https://en.wikipedia.org/wiki/Breadth-first_search">breadth-first search
(BFS)</a> or
<a class="reference external" href="https://en.wikipedia.org/wiki/Depth-first_search">depth-first search
(DFS)</a>.
Optionally, a chunk size specifies how much of the graph to grab at
once in each topological sorting phase. The sorting algorithm will
continue the sort from the next node in the prior graph order that
has not been sorted, yet. The cycle breaking algorithm applies a DFS
sort until a cycle is found. We break and start a new DFS sort phase
from where we stopped.</p></li>
<li><p>A random sort: The graph is randomly sorted. The node order is
randomly shuffled from <a class="reference external" href="http://www.cplusplus.com/reference/random/mt19937/">Mersenne Twister
pseudo-random</a>
generated numbers.</p></li>
<li><p>A sparse matrix mondriaan sort: We can partition a hypergraph with
integer weights and uniform hyperedge costs using the
<a class="reference external" href="http://www.staff.science.uu.nl/~bisse101/Mondriaan/">Mondriaan</a>
partitioner.</p></li>
<li><p>A 1D linear SGD sort: Odgi implements a 1D linear, variation graph
adjusted, multi-threaded version of the <a class="reference external" href="https://arxiv.org/abs/1710.04626">Graph Drawing by Stochastic
Gradient Descent</a> algorithm. The
force-directed graph drawing algorithm minimizes the graph’s energy
function or stress level. It applies stochastic gradient descent
(SGD) to move a single pair of nodes at a time.</p></li>
<li><p>An eades algorithmic sort: Use <a class="reference external" href="http://www.it.usyd.edu.au/~pead6616/old_spring_paper.pdf">Peter Eades’ heuristic for graph
drawing</a>.</p></li>
</ul>
<p>Sorting the paths in a graph my refine the sorting process. For the
users’ convenience, it is possible to specify a whole pipeline of sorts
within one parameter.</p>
<div class="line-block">
<div class="line"><strong>odgi view</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<em>OPTION</em>]…</div>
<div class="line">The odgi view(1) command can convert a graph in odgi format to GFAv1.
It can reveal a graph’s internal structures for e.g. debugging
processes.</div>
</div>
<div class="line-block">
<div class="line"><strong>odgi kmers</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<strong>-c, –stdout</strong>] [<em>OPTION</em>]…</div>
<div class="line">Given a kmer length, the odgi kmers(1) command can emit all kmers. The
output can be refined by setting the maximum number of furcations at
edges or by not considering nodes above a given node degree limit.</div>
</div>
<div class="line-block">
<div class="line"><strong>odgi unitig</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<em>OPTION</em>]…</div>
<div class="line">The odgi unitig(1) command can print all unitigs of a given odgi graph
to standard output in FASTA format. Unitigs can also be emitted in a
fixed sequence quality FASTQ format. Various parameters can refine the
unitigs to print.</div>
</div>
<div class="line-block">
<div class="line"><strong>odgi viz</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<strong>-o, –out</strong>=<em>FILE</em>]
[<em>OPTION</em>]…</div>
<div class="line">The odgi viz(1) command can produce a linear, static visualization of
an odgi variation graph. It aggregates the pangenome into bins and
directly renders a raster image. The binning level depends on the
target width of the PNG to emit. Can be used to produce visualizations
for gigabase scale pangenomes. For more information about the binning
process, please refer to <a class="reference external" href="#odgi_bin.adoc#_odgi_bin1">odgi bin</a>. If
reverse coloring was selected, only the bins with a reverse rate of at
least 0.5 are colored. Currently, there is no parameter to color
according to the sequence coverage in bins available.</div>
</div>
<p><strong>odgi draw</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<strong>-c, –coords-in</strong>=<em>FILE</em>]
[<strong>-p, –png</strong>=<em>FILE</em>] [<em>OPTION</em>]… The odgi draw(1) command draws
previously-determined 2D layouts of the graph with diverse annotations.</p>
<div class="line-block">
<div class="line"><strong>odgi paths</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<em>OPTION</em>]…</div>
<div class="line">The odgi paths(1) command allows the investigation of paths of a given
variation graph. It can calculate overlap statistics of groupings of
paths.</div>
</div>
<div class="line-block">
<div class="line"><strong>odgi prune</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<strong>-o, –out</strong>=<em>FILE</em>]
[<em>OPTION</em>]…</div>
<div class="line">The odgi prune(1) command can remove complex parts of a graph. One can
drop paths, nodes by a certain kind of edge coverage, edges and graph
tips. Specifying a kmer length and a maximum number of furcations, the
graph can be broken at edges not fitting into these conditions.</div>
</div>
<div class="line-block">
<div class="line"><strong>odgi unchop</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<strong>-o, –out</strong>=<em>FILE</em>]
[<em>OPTION</em>]…</div>
<div class="line">The odgi unchop(1) command merges each unitig into a single node.</div>
</div>
<div class="line-block">
<div class="line"><strong>odgi normalize</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<strong>-o, –out</strong>=<em>FILE</em>]
[<em>OPTION</em>]…</div>
<div class="line">The odgi normalize(1) command
<a class="reference external" href="#odgi_unchop.adoc#_odgi_unchop1">unchops</a> a given variation graph
and simplifies redundant furcations.</div>
</div>
<div class="line-block">
<div class="line"><strong>odgi matrix</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<em>OPTION</em>]…</div>
<div class="line">The odgi matrix(1) command generates a sparse matrix format out of the
graph topology of a given variation graph.</div>
</div>
<div class="line-block">
<div class="line"><strong>odgi bin</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<em>OPTION</em>]…</div>
<div class="line">The odgi bin(1) command bins a given variation graph. The pangenome
sequence, the one-time traversal of all nodes from smallest to largest
node identifier, can be summed up into bins of a specified size. For
each bin, the path metainformation is summarized. This enables a
summarized view of gigabase scale graphs. Each step of a path is a bin
and connected to its next bin via a link. A link has a start bin
identifier and an end bin identifier.</div>
<div class="line">The concept of odgi bin is also applied in odgi
<a class="reference external" href="#odgi_viz.adoc#_odgi_viz1">viz</a>. A demonstration of how the odgi
bin JSON output can be used for an interactive visualization is
realized in the <a class="reference external" href="https://graph-genome.github.io/">Pantograph</a>
project. Per default, odgi bin writes the bins to stdout in a
tab-delimited format: <strong>path.name</strong>, <strong>path.prefix</strong>, <strong>path.suffix</strong>,
<strong>bin</strong> (bin identifier), <strong>mean.cov</strong> (mean coverage of the path in
this bin), <strong>mean.inv</strong> (mean inversion rate of this path in this
bin), <strong>mean.pos</strong> (mean nucleotide position of this path in this
bin), <strong>first.nucl</strong> (first nucleotide position of this path in this
bin), <strong>last.nucl</strong> (last nucleotide position of this path in this
bin). These nucleotide ranges might span positions that are not
present in the bin. Example: A range of 1-100 means that the first
nucleotide has position 1 and the last has position 100, but
nucleotide 45 could be located in another bin. For an exact positional
output, please specify [<strong>-j, –json</strong>].</div>
</div>
<div class="line-block">
<div class="line"><strong>odgi chop</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<strong>-o, –out</strong>=<em>FILE</em>]
[<strong>-c, –chop-to</strong>=<em>N</em>] [<em>OPTION</em>]…</div>
<div class="line">The odgi chop(1) command chops long nodes into short ones while
preserving the graph topology.</div>
</div>
<div class="line-block">
<div class="line"><strong>odgi layout</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<strong>-o, –out</strong>=<em>FILE</em>]
[<em>OPTION</em>]…</div>
<div class="line">The odgi layout(1) command computes 2D layouts of the graph using
stochastic gradient descent (SGD). The input graph must be sorted and
id-compacted. The algorithm itself is described in <a class="reference external" href="https://arxiv.org/abs/1710.04626">Graph Drawing by
Stochastic Gradient Descent</a>. The
force-directed graph drawing algorithm minimizes the graph’s energy
function or stress level. It applies SGD to move a single pair of
nodes at a time.</div>
</div>
<div class="line-block">
<div class="line"><strong>odgi flatten</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<em>OPTION</em>]…</div>
<div class="line">The odgi flatten(1) command projects the graph sequence and paths into
FASTA and BED.</div>
</div>
<div class="line-block">
<div class="line"><strong>odgi break</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<strong>-o, –out</strong>=<em>FILE</em>]
[<em>OPTION</em>]…</div>
<div class="line">The odgi break(1) command finds cycles in a graph via <a class="reference external" href="https://en.wikipedia.org/wiki/Breadth-first_search">breadth-first
search (BFS)</a>
and breaks them, also dropping the graph’s paths.</div>
</div>
<div class="line-block">
<div class="line"><strong>odgi pathindex</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<strong>-o, –out</strong>=<em>FILE</em>]
[<em>OPTION</em>]…</div>
<div class="line">The odgi pathindex(1) command generates a path index of a graph. It
uses succinct data structures to encode the index. The path index
represents a subset of the features of a fully realized <a class="reference external" href="https://github.com/vgteam/xg">xg
index</a>. Having a path index, we can
use odgi <a class="reference external" href="#odgi_panpos.adoc#_odgi_panpos1">panpos</a> to go from
<strong>path:position</strong> → <strong>pangenome:position</strong> which is important when
navigating large graphs in an interactive manner like in the
<a class="reference external" href="https://graph-genome.github.io/">Pantograph</a> project.</div>
</div>
<div class="line-block">
<div class="line"><strong>odgi panpos</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<strong>-p, –path</strong>=<em>STRING</em>]
[<strong>-n, –nuc-pos</strong>=<em>N</em>] [<em>OPTION</em>]…</div>
<div class="line">The odgi panpos(1) command give a pangenome position for a given path
and nucleotide position. It requires a path index, which can be
created with odgi
<a class="reference external" href="#odgi_pathindex.adoc#_odgi_pathindex1">pathindex</a>. Going from
<strong>path:position</strong> → <strong>pangenome:position</strong> is important when
navigating large graphs in an interactive manner like in the
<a class="reference external" href="https://graph-genome.github.io/">Pantograph</a> project. All input
and output positions are 1-based.</div>
</div>
<div class="line-block">
<div class="line"><strong>odgi server</strong> [<strong>-i, –idx</strong>=<em>FILE</em>] [<strong>-p, –port</strong>=<em>N</em>]
[<em>OPTION</em>]…</div>
<div class="line">The odgi server(1) command starts an HTTP server with a given path
index as input. The idea is that we can go from <strong>path:position</strong> →
<strong>pangenome:position</strong> via GET requests to the HTTP server. The server
headers do not block cross origin requests. Example GET request:
<em>http://localost:3000/path_name/nucleotide_position</em>.</div>
<div class="line">The required path index can be created with odgi
<a class="reference external" href="#odgi_pathindex.adoc#_odgi_pathindex1">pathindex</a>. Going from
<strong>path:position</strong> → <strong>pangenome:position</strong> is important when
navigating large graphs in an interactive manner like in the
<a class="reference external" href="https://graph-genome.github.io/">Pantograph</a> project. All input
and output positions are 1-based. If no IP address is specified, the
server will run on localhost.</div>
</div>
<div class="line-block">
<div class="line"><strong>odgi test</strong> [&lt;TEST NAME|PATTERN|TAGS&gt; …] [<em>OPTION</em>]…</div>
<div class="line">The odgi test(1) command starts all unit tests that are implemented in
odgi. For targeted testing, a subset of tests can be selected. odgi
test(1) depends on <a class="reference external" href="https://github.com/catchorg/Catch2">Catch2</a>. In
the default setting, all results are printed to stdout.</div>
</div>
<div class="line-block">
<div class="line"><strong>odgi version</strong> [<em>OPTION</em>]…</div>
<div class="line">The odgi version(1) command prints the current git version with tags
and codename to stdout (like <em>v-44-g89d022b “back to old ABI”</em>).
Optionally, only the release, version or codename can be printed.</div>
</div>
</div>
<div class="section" id="bugs">
<h1>BUGS<a class="headerlink" href="#bugs" title="Permalink to this headline">¶</a></h1>
<p>Refer to the <strong>odgi</strong> issue tracker at
<a class="reference external" href="https://github.com/pangenome/odgi/issues">https://github.com/pangenome/odgi/issues</a>.</p>
</div>
<div class="section" id="authors">
<h1>AUTHORS<a class="headerlink" href="#authors" title="Permalink to this headline">¶</a></h1>
<p>Erik Garrison from the University of California Santa Cruz wrote the
whole <strong>odgi</strong> tool. Simon Heumos from the Quantitative Biology Center
Tübingen wrote <strong>odgi pathindex</strong>, <strong>odgi panpos</strong>, <strong>odgi server</strong>, and
this documentation. Andrea Guarracino from the University of Rome Tor
Vergata wrote <strong>odgi viz</strong>, <strong>odgi extract</strong>, <strong>odgi cover</strong>, <strong>odgi
explode</strong>, <strong>odgi squeeze</strong>, <strong>odgi depth</strong>, <strong>odgi overlap</strong>, <strong>odgi
validate</strong>, and this documentation.</p>
</div>
<div class="section" id="resources">
<h1>RESOURCES<a class="headerlink" href="#resources" title="Permalink to this headline">¶</a></h1>
<p><strong>Project web site:</strong> <a class="reference external" href="https://github.com/pangenome/odgi">https://github.com/pangenome/odgi</a></p>
<p><strong>Git source repository on GitHub:</strong> <a class="reference external" href="https://github.com/pangenome/odgi">https://github.com/pangenome/odgi</a></p>
<p><strong>GitHub organization:</strong> <a class="reference external" href="https://github.com/pangenome">https://github.com/pangenome</a></p>
<p><strong>Discussion list / forum:</strong> <a class="reference external" href="https://github.com/pangenome/odgi/issues">https://github.com/pangenome/odgi/issues</a></p>
</div>
<div class="section" id="copying">
<h1>COPYING<a class="headerlink" href="#copying" title="Permalink to this headline">¶</a></h1>
<p>The MIT License (MIT)</p>
<p>Copyright (c) 2019-2021 Erik Garrison</p>
<p>Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the
“Software”), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:</p>
<p>The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.</p>
<p>THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.</p>
</div>


           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Erik Garrison. Revision v0.5.1-e809b89.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>